### PR TITLE
Search reference in all combinations

### LINKED ISSUE #9727: [BOOM-4928] Listings (stock & products) can't search by product combination reference
> This issue has been migrated from the Forge. Read the [original ticket here](http://forge.prestashop.com/browse/BOOM-4928).

- _**Reporter:**_ thbl
- _**Created at:**_ Sat, 24 Feb 2018 19:08:47 +0100

If you search by a product combination's reference in admin stocks you get no result.
Tho if you look on the stock list "without using search" you can see the combination's  and reference

---
#### Describe the bug 

I have products with combinations. The product itself and each one of the combination have a ref code.
If i search a product by it's ref, in the stock page i'm only able to find the original product, not the combinations.
In the stock movement page, i'm able to make it work.

#### Expected behavior and what happen

Example: i have a product named "slippers" with a ref code "abc". There are 4 combinations for this product: 1: n°36 (ref code "1"), 2: n°37 (ref code "2"), 3: n°38 (ref code "3"), 4: n°39 (ref code "4").

In the stock page, if i search for "abc", i find all the combinations. If i search for "1" i find no results. Idem if i search for "abc 1".

In the stock movement page i have these results: for "abc" i find all the combinations. for "1" i find the combination "n°36", same result if i search for "abc 1".

Why the stock page search doesn't work like the stock movement one?

#### Steps to reproduce 

1. 

2. 
3. 
4. 
5. 
6. 
7. 
8. 
9. 
10. 
11. 
12. 

#### Additional information

* PrestaShop version: 1.7.8.0
* PHP version: 7.3.26

### LINKED ISSUE #20443: BO - Pack can't search by product combination reference
#### Describe the bug

I have products, which have combinations that have their own reference. I want to make product packs that include combinations.

When I try to update reference in combination, for example from "A1234" to "A1234 F12", product pack search does not refresh and I cannot search by this new piece of reference. FO search works fine.

#### Expected behavior

When I update combination reference and try to search combination into product pack, I should be able to search by updated reference.

Steps to reproduce the issue

1. Create a product with combinations Product1
2. Set for the first combination reference =A1234
3. Set for the first combination reference =A1234 B12
4. Create a new pack, try to search for the last product created with `A1234` OR `B12`
5. No result
6. Go back to the product Product1and update the default reference with "A1234 B12"
7. Create a pack and try to search with A1234 or B12 => OK

----------------------------------------------

1. Create product
2. Create combination to product
3. Add reference "A1234" to combination and save
4. Go edit the combination and edit reference to "A1234 B34"
5. Create a new product pack
6. Try to search "A12" - product is shown
7. Try to search "b34" - nothing is shown

**Screenshots**

Working search:

product-search-works

Nothing is found:

product-search-not-working

#### Additional information

I have tried to:
- clear cache
- remove cache manually from /var/cache
- rebuilt product index

But nothing has helped.

* PrestaShop version: 1.7.6.5
* PHP version: 7.3

### LINKED ISSUE #20454: If ID_product is not defined, we can't find combination by reference
### Describe the bug

If there is one row in SQL table product_attribute that has id_product = 25 and reference = demo_10

Then if I run `$id = Combination::getIdByReference(25, 'demo_1');` this will return the above record because a LIKE is used in the query instead of a equality.

`$id = Combination::getIdByReference(25, 'demo_1');` should return a result if there is a row that has id_product = 25 and reference = demo_1 not demo_10

<hr>

#### Describe the bug

Actually, we can find a combination by reference only related to a specific product.
We could be retrieve combination in global context without take in consideration a specific product.

> Reported [here](https://github.com/PrestaShop/PrestaShop/issues/26467) by @saiif

Combination::getIdByReference() returns wrong id 

#### Expected behavior

$id = Combination::getIdByReference(500);
return the same id 
$id = Combination::getIdByReference(5000);
cause like is used for the query 

#### Steps to Reproduce

$id = Combination::getIdByReference(500);
return the same id 
$id = Combination::getIdByReference(5000);

-------------------------------------------------------------------------

#### Expected behavior

Search reference without provide a product information
You can follow this [comment](https://github.com/PrestaShop/PrestaShop/issues/20454#issuecomment-668689318)
#### Steps to Reproduce

Not relevant

#### Additional information

* PrestaShop version: develop
* PHP version: 7.x

### PR DESCRIPTION
**Pull request by camlafit**
solve #20454

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | * Prevent a PHP error when idproduct is not set<br>* Allow search in all combinations, when product is unknown
| Type?         | bug fix 
| Category?     | CO 
| BC breaks?    |  no
| Deprecations? |  no
| Fixed ticket? | Fixes #20454
| How to test?  | Do a code as Combination::getIdByReference(0,1234) or Combination::getIdByReference('',1234), <br> Thanks to check if those issues are fixed: https://github.com/PrestaShop/PrestaShop/issues/20443 & https://github.com/PrestaShop/PrestaShop/issues/9727

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/20455)

### CODE REVIEW

**Comment by ghost** [comment]
Hello

A little up about this PR. 
thanks

**Review by Progi1984 — DISMISSED** [review]

File: classes/Combination.php [hunk 1]
```diff
@@ -408,6 +408,7 @@ public static function isCurrentlyUsed($table = null, $hasActiveColumn = false)
 
     /**
      * For a given product_attribute reference, returns the corresponding id.
+     * if idProduct is empty|0 search is done in all combinations
      *
      * @param int $idProduct
```
Comment (Progi1984) [inline classes/Combination.php]:
```suggestion
     * @param int|null $idProduct
```

File: classes/Combination.php [hunk 2]
```diff
@@ -408,6 +408,7 @@ public static function isCurrentlyUsed($table = null, $hasActiveColumn = false)
 
     /**
      * For a given product_attribute reference, returns the corresponding id.
+     * if idProduct is empty|0 search is done in all combinations
```
Comment (Progi1984) [inline classes/Combination.php]:
```suggestion
     * If idProduct is empty(0/null), search is done in all combinations
```

Comment (ghost) [reply classes/Combination.php]: done

File: classes/Combination.php (hunk 1 above)
Comment (ghost) [reply classes/Combination.php]: done

**Review by ghost — APPROVED** [review]

**Review by Progi1984 — APPROVED** [review]

File: classes/Combination.php [hunk 3]
```diff
@@ -424,7 +425,9 @@ public static function getIdByReference($idProduct, $reference)
         $query->select('pa.id_product_attribute');
         $query->from('product_attribute', 'pa');
         $query->where('pa.reference LIKE \'%' . pSQL($reference) . '%\'');
-        $query->where('pa.id_product = ' . (int) $idProduct);
+        if (!empty($idProduct)) {
+            $query->where('pa.id_product = ' . (int) $idProduct);
```
Comment (PierreRambaud) [inline classes/Combination.php:429]: The reference is not unique can returns an array.
The reference use like, it's not a good idea.
What you probably want is to add a new function
```php
    /**
     * For a given product_attribute reference, returns the product ids.
     *
     * @param string $reference
     *
     * @return int id
     */
    public static function getByReference($reference)
    {
        if (empty($reference)) {
            return [];
        }

        $query = new DbQuery();
        $query->select('pa.id_product_attribute, pa.id_product');
        $query->from('product_attribute', 'pa');
        $query->where('pa.reference = \'' . pSQL($reference) . '\'');

        return Db::getInstance(_PS_USE_SQL_SLAVE_)->executeS($query);
    }
```

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
Not sure it's the right way to do what you want.

File: classes/Combination.php (hunk 3 above)
Comment (camlafit) [reply classes/Combination.php:429]: Hello

I can rework this code but these last remarks don't look related to this PR. Theses notes concern yet the actual code ( https://github.com/PrestaShop/PrestaShop/blob/develop/classes/Combination.php#L442-L455) . 
Like is yet used on query and we can have many reference with this query as any unique index are set on id_product/reference.

Do you confirm the need of this code rework ?

Comment (PierreRambaud) [reply classes/Combination.php:429]: Yes, I confirm it needs to be refactor what you want is something a little different than the original function (which is badly named).

Comment (camlafit) [reply classes/Combination.php:429]: Hello

Understood. I rework on this.

**Comment by PierreRambaud** [comment]
Hi @camlafit ,

Since we had no news from you for more than 30 days, I'll close this pull request. Feel free to reopen or open another one if you think it's still relevant.

Thanks!

**Comment by camlafit** [comment]
Hello

Sorry but I don't understand some process/requirement here. #20454 was waiting response from prestashop team not me.
For me it's more in more complicate to try to contribute. I've tried to follow all requirement, open issue, explain it, provide solution, don't break any existing code, but looks not enougth but I don't understand why.

**Pull request closed without merging**